JPS_BASE_LIB/src/main/java/uk/ac/cam/cares/jps/base/email/EmailSender.java

```diff
@@ -0,0 +1,262 @@
+package uk.ac.cam.cares.jps.base.email;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetAddress;
+import java.net.URL;
+import java.net.UnknownHostException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Optional;
+import java.util.Properties;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import uk.ac.cam.cares.jps.base.discovery.AgentCaller;
+
+/**
+ * This class handles sending HTTP requests to a remote EmailAgent instance to facilitate the
+ * sending of an automated email.
+ *
+ * @author Michael Hillman
+ */
+public class EmailSender {
+
+    /**
+     * Fallback URL for the remote email agent.
+     */
+    private static final String DEFAULT_AGENT_URL = "http://kg.cmclinnovations.com/agents/email-agent";
+
+    /**
+     * Contains meta-data to append to email content.
+     */
+    private final Properties metadata;
+
+    /**
+     * URL for remote EmailAgent instance.
+     */
+    private String emailAgentURL;
+
+    /**
+     * Denotes if the remote EmailAgent is reachable and available.
+     */
+    private boolean isAgentReachable = false;
+
+    /**
+     * Create a new EmailSender instance.
+     */
+    public EmailSender() {
+        this.metadata = new Properties();
+
+        // Can we reach the remote EmailAgent?
+        determineAgentLocation();
+        this.isAgentReachable = isReachable();
+
+        // Generate meta-data to add to email
+        gatherMetaData();
+    }
+
+    /**
+     * Attempts to send a HTTP request to the remote EmailAgent instance using the input content.
+     *
+     * Note: if this machine is not running on the same network as the EmailAgent instance (or the
+     * remote EmailAgent cannot be reached), then the same content is written to a local file
+     * instead.
+     *
+     * @param subject email subject
+     * @param body email body
+     *
+     * @return Optionally returns location of log file (if written to).
+     *
+     * @throws Exception if not initialised, email cannot be sent or written to local file.
+     */
+    public Optional<Path> sendEmail(String subject, String body) throws Exception {
+        // Append the meta-data to the email body
+        StringBuilder strBuilder = new StringBuilder(body);
+        strBuilder.append("<br><br>");
+        strBuilder.append("The following meta-data was gathered from the machine triggering this notification:");
+        strBuilder.append("<br>");
+
+        for (Object key : metadata.keySet()) {
+            strBuilder.append("&#9;" + key.toString() + ": " + metadata.get(key).toString());
```

Comment:
Best to avoid using `+` within `append` and don't need to use `toString` explicitly with either.
```suggestion
            strBuilder.append("&#9;").append(key).append(": ").append(metadata.get(key));
```